packages/core/src/types.ts

```diff
@@ -321,28 +321,34 @@ export type StatesDefinition<
   >;
 };
 
-export type TransitionConfigTargetShortcut<
+export type TransitionConfigTarget<TContext, TEvent extends EventObject> =
+  | string
+  | undefined
+  | StateNode<TContext, any, TEvent>;
+
+export type TransitionConfigOrTarget<
   TContext,
   TEvent extends EventObject
-> = string | undefined | StateNode<TContext, any, TEvent>;
+> = SingleOrArray<
+  TransitionConfigTarget<TContext, TEvent> | TransitionConfig<TContext, TEvent>
+>;
 
 type TransitionsConfigMap<TContext, TEvent extends EventObject> = {
-  [K in TEvent['type'] | NullEvent['type'] | '*']?: SingleOrArray<
-    | TransitionConfigTargetShortcut<TContext, TEvent>
-    | (TransitionConfig<
-        TContext,
-        K extends TEvent['type'] ? Extract<TEvent, { type: K }> : EventObject
-      > & {
-        event?: undefined;
-      })
+  [K in TEvent['type']]?: TransitionConfigOrTarget<
+    TContext,
+    TEvent extends { type: K } ? TEvent : never
   >;
+} & {
+  ''?: TransitionConfigOrTarget<TContext, TEvent>;
+} & {
+  '*'?: TransitionConfigOrTarget<TContext, TEvent>;
 };
 
 type TransitionsConfigArray<TContext, TEvent extends EventObject> = Array<
   {
     [K in TEvent['type'] | NullEvent['type'] | '*']: TransitionConfig<
       TContext,
-      K extends TEvent['type'] ? Extract<TEvent, { type: K }> : EventObject
+      TEvent extends { type: K } ? TEvent : never
```

Comment:
is the following `& { event: K }` needed now? it might be obsolete, but im not 100% sure

Reply:
this probably was already the case but this type doesn't really allow for null and wildcard events to be configured here because they will be mapped to `never` (they dont extend TEvent after all)

Reply:
Yes, `{ event: K }` is needed to narrow array transition configs of this style, as you have added:

```js
on: [
  { event: 'something', ... }
  { event: 'another', ... }
]
```